### PR TITLE
hubble: node labels

### PR DESCRIPTION
**Pull request by kaworu**
This PR adds node labels info to Hubble flows along with a corresponding `--node-label` filter.

Filtering flows emitted from nodes having particular labels can be useful to debug the Egress Gateway feature: combined with the recently added network interface filter and/or SNAT IP filter one could then see egress flows related to a given `CiliumEgressGatewayPolicy`.

### CODE REVIEW

File: pkg/hubble/filters/labels.go [hunk 1]
```diff
@@ -90,5 +95,13 @@ func (l *LabelsFilter) OnBuildFilter(ctx context.Context, ff *flowpb.FlowFilter)
 		fs = append(fs, dlf)
 	}
 
+	if ff.GetNodeLabels() != nil {
+		nlf, err := FilterByLabelSelectors(ff.GetNodeLabels(), nodeLabels)
+		if err != nil {
+			return nil, fmt.Errorf("invalid node label filter: %v", err)
```
Comment (gandro) [inline pkg/hubble/filters/labels.go]: Please use `%w` here

**Review by gandro — APPROVED** [review]
One minor nit, but otherwise this looks good to me!

File: pkg/hubble/observer/local_node_watcher.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	w.updated++
+	w.mu.Unlock()
+}
+
+// generation return this watcher update counter.
+func (w *LocalNodeWatcher) generation() int {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.updated
+}
```
Comment (michi-covalent) [inline pkg/hubble/observer/local_node_watcher.go:62]: does this function get used somewhere?

**Review by michi-covalent — APPROVED** [review]
lgtm one minor question.

File: pkg/hubble/filters/labels.go (hunk 1 above)
Comment (kaworu) [reply pkg/hubble/filters/labels.go]: Nice catch, done! :pray:

File: pkg/hubble/observer/local_node_watcher.go (hunk 2 above)
Comment (kaworu) [reply pkg/hubble/observer/local_node_watcher.go:62]: Yes, it is used in the test.

In `NewLocalNodeWatcher`, observing the local node is setup but we get the first emit/update event asynchronously. In other words, the local node watcher cache is probably empty when `NewLocalNodeWatcher` returns. Consequently, if we call `NewLocalNodeWatcher` and check its cache right away we get inconsistent results, so I added a cache update count and the test check it like this:
```go
require.Eventually(t, func() bool {
	return watcher.generation() > 0
}, time.Second, 10*time.Millisecond)
```
The `generation()` func is a helper ensuring to hold the lock.

**Review by chancez — APPROVED** [review]

**Comment by michi-covalent** [comment]
looks like we need to update hubble observe help text: https://github.com/cilium/cilium/actions/runs/9512851249/job/26221795069
```
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -68,2 +68,3 @@
            	            	   -n, --namespace filter                    Show all flows related to the given Kubernetes namespace.
            	            	+      --node-label filter                   Show only flows observed on nodes matching the given label filter (e.g. "key1=value1", "io.cilium/egress-gateway")
            	            	       --node-name filter                    Show all flows which match the given node names (e.g. "k8s*", "test-cluster/*.company.com")
            	Test:       	TestTestHubbleObserve/observe_help
            	Messages:   	expected output does not match
```

**Pull request merged**